Thermal properties of fluids, including refrigerants, air, and water, etc., are very commonly used in various analytics efforts. This article introduces the thermal properties module in the UTC Data Analytics Library.

Currently, the integrated thermal properties module is based on the open source [Coolprop](http://www.coolprop.org/) package, which is a C++ library that integrates the equation of state of over 100 pure or pseudo-pure fluids.

> *Warning*: Although the Coolprop library is widely used and most [data sources](http://www.coolprop.org/fluid_properties/PurePseudoPure.html#list-of-fluids) should be reliable, it cannot be fully trusted if you are to do some serious calculations. The bottom line is, at least perform some kind of cross check within the interested range, with *standard* data source such as Refprop from NIST.

# Basic usage

Import the `props_SI` function as:

In [1]:
from UTCDAL.Utilities.Domain.Thermodynamics.thermal_props import props_SI

There is also a shortcut for it if you feel the import chain is a bit too long:

```python
from UTCDAL.namespaces import props_SI
```

In [2]:
help(props_SI)

Help on function props_SI in module UTCDAL.Utilities.Domain.Thermodynamics.thermal_props:

props_SI(*args, **kwargs)
    High-level function for calculating thermal properties of fluids.
    
    Sample use 1 (coolprop-like interface, with 6 arguments):
    >>> import UTCDAL.namespaces as dal
    >>> # calculate single state point
    >>> # arguments should be ordered as follows
    >>> dal.props_SI('C', 'T', 300, 'P', 101325, 'R410a')
    >>> # calculate multiple state points by giving some iterable (list, etc)
    >>> dal.props_SI('C', 'T', range(300, 350), 'P', 101325, 'R410a')
    
    Sample use 2 (with 4 arguments):
    >>> dal.props_SI('C', 'R134a', T=300, P=101325)
    
    Sample use 3 (with 2 arguments):
    >>> dal.props_SI('Tcrit', 'Water')
    
    To check available fluids, call without argument:
    >>> dal.props_SI()



# Available fluids

All available fluids in the current version are:

In [3]:
fluids = props_SI()

fluids[:6]

['1-Butene', 'Acetone', 'Air', 'Ammonia', 'Argon', 'Benzene']

In [4]:
len(fluids)

122

There are in total 122 kinds of fluids in the database.

# Example usage of `props_SI` function

The first way of using `props_SI` function is created only for compatibility reasons. It is the default way if you are using the coolprop's Python binding, and similar to Refprop's default binding as well. The recommended ways are the last two ones, for instance:

```python
props_SI('C', 'R134a', T=300, P=101325)
props_SI('Tcrit', 'Water')
```

The syntax is more concise and more readable as well.

* Density of carbon dioxide at 100 bar and 25C

In [5]:
props_SI('D', 'CO2', T=298.15, P=100e5)

817.6273812375758

* Saturated vapor enthalpy [J/kg] of R134a at 25C

In [6]:
props_SI('H', 'R134a', T=298.15, Q=1)

412333.95323186804

* Boiling temperature of water at 101325 Pa

In [7]:
props_SI('T', 'Water', P=101325, Q=0)

373.1242958476844

The above function calls can all accept inputs as iterables like list. It could be handy for analysis.

In [8]:
import numpy as np

In [9]:
props_SI('T', 'Water', P=np.linspace(0.9e5, 1.2e5, 5), Q=0)

[369.83714755587414,
 372.04974504589956,
 374.12552064990285,
 376.08202331372576,
 377.93354962911144]

Notice that we have been using shorthands, e.g. T for temperature, D for density. You can find the full table from [coolprop's online reference](http://www.coolprop.org/coolprop/HighLevelAPI.html#parameter-table). We also included a table of commonly used variables:

| Parameter | Description |
| --------- | ----------- |
| D | Density |
| H | Mass specific enthalpy |
| P | Pressure |
| Q | Mass vapor quality |
| S | Mass specific entropy |
| T | Temperature |
| U | Mass specific internal energy |
| A | Speed of sound |
| C | Mass specific constant pressure specific heat |
| I | Surface tension |
| V | Viscosity |